### ISSUE TITLE
Signed Integer Overflow in MoE _unpermute with Inductor Compilation

### ISSUE THREAD

**Issue by hirschsn** [pytorch/torchtitan#2430]
### Bug description

tl;dr: There is an integer overflow in the Torch Inductor Triton implementation of `aten::index_put` in the MoE grouped_experts unpermute.

# Details

If you enable Torch Inductor compilation for the MoE grouped GEMMs (needs a code change, not only `enable=True`, c.f. https://github.com/pytorch/torchtitan/pull/2322) and run a large model, I get:

- Silent memory corruption (e.g. we observed loss and grad norm NaNs)
- Illegal memory access (e.g. reported by `cudaGetLastError`), training crashes

Backtrace:

- Triton kernel for `aten::index_put`
- `_unpermute` (torchtitan/models/moe/utils.py:65)
- `indices_padding_wrapper` (torchtitan/models/moe/utils.py:93)
- `GroupedExperts.forward`(torchtitan/models/moe/moe.py:175)
- `MoE.forward` (torchtitan/models/moe/moe.py:524)

The calling code in `_unpermute`:

```
    out_unpermuted[permuted_indices, :] = out
```

Triton implementation emitted by the Torch Inductor:

```
def triton_poi_fused_index_put_10(permuted_indices, out, out_unpermuted, ...):
    # ... omitted for brevity ...
    # tmp6 is a value from "out" to be copied to "out_unpermuted"
    # tmp4 is an int32 index into the "out_unpermuted" tensor (comes from permuted_indices)
    tl.device_assert((0 <= tmp4) & (tmp4 < 1048577), "index out of bounds: 0 <= tmp4 < 1048577")
    tl.store(out_unpermuted + (x0 + 2048*tmp4), tmp6, None) # <--- possible overflow 2048*tmp4
```

These indices are all 32-bit due to `permuted_indices` being a Tensor with 32-bit indices.
 
- The 32-bit indices are fine here, since they are in `{0, ..., 1048576}`
- But to index into the output tensor, this value gets multiplied with "out_unpermute"'s stride (=2048), which causes an overflow.

The int32 Tensor comes from:

- `fill_index_wrapper` (torchtitan/models/moe/kernels.py:80)
- `generate_permute_indices` (torchtitan/models/moe/kernels.py:207)
- `_permute` (torchtitan/models/moe/utils.py:47)
- `indices_padding_wrapper` (torchtitan/models/moe/utils.py:87)
- ... see backtrace above ...

This tensor should be changed to dtype int64, which would make it consistent with other PyTorch operations that return tensors of indices.

### Versions

Torch Titan HEAD, also with PyTorch nightly.
